lib_comfyui/polling_client.py

```diff
@@ -0,0 +1,164 @@
+import asyncio
+import atexit
+import multiprocessing
+import queue
+
+from lib_comfyui.parallel_utils import clear_queue, StoppableThread
+from lib_comfyui import ipc, global_state
+from lib_comfyui.queue_tracker import PromptQueueTracker
+
+
+# rest is ran on comfyui's server side
+class ComfyuiNodeWidgetRequests:
+    start_comfyui_queue = multiprocessing.Queue()
+    finished_comfyui_queue = multiprocessing.Queue()
+    cids = {}
+    param_events = {}
+    last_params = None
+    loop = None
+    focused_key = None
+
+    @ipc.confine_to('comfyui')
+    @staticmethod
+    def send(request_params):
+        clear_queue(ComfyuiNodeWidgetRequests.finished_comfyui_queue)
+        ComfyuiNodeWidgetRequests.start_comfyui_queue.put(request_params)
+        return ComfyuiNodeWidgetRequests.finished_comfyui_queue.get()
+
+    @ipc.confine_to('comfyui')
+    @staticmethod
+    def start_workflow_sync(batch, workflow_type, is_img2img, required_node_types, queue_front):
+        xxx2img = ("img2img" if is_img2img else "txt2img")
+        setattr(global_state, f'{xxx2img}_{workflow_type}_input_images', batch)
+        setattr(global_state, f'tab_name', xxx2img)
+        PromptQueueTracker.update_tracked_id()
+        response = ComfyuiNodeWidgetRequests.send({
+            'request': '/sd-webui-comfyui/webui_request_queue_prompt',
+            'workflowType': f'comfyui_{workflow_type}_{xxx2img}',
+            'requiredNodeTypes': required_node_types,
+            'queueFront': queue_front,
+        })
+
+        if 'error' in response:
+            return response
+
+        # unsafe queue tracking
+        PromptQueueTracker.wait_for_last_put()
+
+        output_key = f'{xxx2img}_{workflow_type}_output_images'
+        results = getattr(global_state, output_key, None)
+
+        if output_key in global_state:
+            delattr(global_state, output_key)
+
+        return results
+
+    @classmethod
+    def set_loop(cls, loop):
+        if loop is None or cls.loop is not None:
+            return
+
+        cls.loop = loop
+        cls._start_listener()
+
+    @classmethod
+    def _start_listener(cls):
+        def request_listener():
+            while request_thread.is_running():
+                try:
+                    cls.last_params = cls.start_comfyui_queue.get(timeout=1)
+                except queue.Empty:
+                    continue
+                key = ComfyuiNodeWidgetRequests.focused_key
+                cls.loop.call_soon_threadsafe(cls.param_events[key][cls.last_params['workflowType']].set)
+
+        def graceful_exit():
+            nonlocal request_thread
+            try:
+                request_listener()
+            except OSError:
+                pass
+
+        request_thread = StoppableThread(target=graceful_exit)
+        request_thread.start()
+        atexit.register(request_thread.join)
+
+    @classmethod
+    def add_client(cls, cid, key):

```

Comment:
Not 100% sure this is desirable, but is it possible to have this function accept a data object with the 2 ids and register the object only? Or because of the way the values below are queried, it should stay like this?

Reply:
Imo this would clutter the code more than make it more readable?

Reply:
I would rather not do that and leave them separated like that if you think that's fine

Reply:
There's some coupling that could be done, idk